packages/block-editor/src/components/url-input/index.js

```diff
@@ -18,233 +18,286 @@ import {
-					this.props.debouncedSpeak(
-						sprintf(
-							/* translators: %s: number of results. */
-							_n(
-								'%d result found, use up and down arrow keys to navigate.',
-								'%d results found, use up and down arrow keys to navigate.',
-								suggestions.length
-							),
-							suggestions.length
-						),
-						'assertive'
-					);
-				} else {
-					this.props.debouncedSpeak(
-						__( 'No results.' ),
-						'assertive'
-					);
-				}
-				this.isUpdatingSuggestions = false;
-			} )
-			.catch( () => {
-				if ( this.suggestionsRequest === request ) {
-					this.setState( {
-						loading: false,
-					} );
-					this.isUpdatingSuggestions = false;
-				}
-			} );
+	function speakUpdateResult() {
+		if ( ! suggestions.length ) {
+			debouncedSpeak( __( 'No results' ), 'assertive' );
+			return;
+		}
 
-		// Note that this assignment is handled *before* the async search request
-		// as a Promise always resolves on the next tick of the event loop.
-		this.suggestionsRequest = request;
+		debouncedSpeak(
+			sprintf(
+				/// translators: %s: number of results.
+				_n(
+					'%d result found, use up and down arrow keys to navigate.',
+					'%d results found, use up and down arrow keys to navigate.',
+					suggestions.length
+				),
+				suggestions.length
+			),
+			'assertive'
+		);
 	}
 
-	onChange( event ) {
-		const inputValue = event.target.value;
+	const updateSuggestions = debounce( async ( search = '' ) => {
+		setSelectedSuggestion( null );
 
-		this.props.onChange( inputValue );
-		if ( ! this.props.disableSuggestions ) {
-			this.updateSuggestions( inputValue.trim() );
+		if ( ! allowSuggestionsRequest( search ) ) {
+			setShowSuggestions( false );
+			setLoading( false );
+			return;
 		}
-	}
+		setLoading( true );
 
-	onFocus() {
-		const { suggestions } = this.state;
-		const { disableSuggestions, value } = this.props;
+		const request = fetchLinkSuggestions( search, {
+			isInitialSuggestions: ! search.length,
+		} );
+		pendingRequest.current = request;
+
+		try {
+			const result = await request;
+			if ( result && request === pendingRequest.current ) {
+				setSuggestions( result );
+				setShowSuggestions( !! result.length );
+				speakUpdateResult();
+				clearUpdateSuggestions();
+			}
+		} catch ( _error ) {
+			// Release updating state if the current request fails.
+			if ( pendingRequest.current === request ) {
+				clearUpdateSuggestions();
+			}
+		}
+	}, REQUEST_DEBOUNCE_DELAY );
 
+	function onFocus() {
 		// When opening the link editor, if there's a value present, we want to load the suggestions pane with the results for this input search value
 		// Don't re-run the suggestions on focus if there are already suggestions present (prevents searching again when tabbing between the input and buttons)
 		if (
 			value &&
 			! disableSuggestions &&
-			! this.isUpdatingSuggestions &&
-			! ( suggestions && suggestions.length )
+			! isUpdatingSuggestions() &&
+			! suggestions.length
```

Comment:
```suggestion
			! suggestions?.length
```

Reply:
Not sure this guard is needed. `suggestions` is initialized to an empty array in a `useState` call.  There is one call to `setSuggestions` with the result from the suggestions fetch but there is an _truthy_ check before it's called. I'll add an `Array.isArray` check on the result (length won't work here since we want to update suggestions if they are empty)

Reply:
Added Array check in d5fe5e812657180407087b3d17c4071d86d27f9d

Reply:
Actually Array.isArray catches null and undefined, so you need just this check without checking the plain result.

Reply:
The conditional is something like `result && Array.isArray( result )` which should prevent the `setSuggestions` call with null or undefined. But I thought `Array.isArray` will return false for null or undefined ? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray#using_array.isarray

Reply:
I'm not sure I understood correctly but my comment was about `result && Array.isArray( result )` being equivalent to `Array.isArray(result)`

Reply:
ah I see, I total misread your comment @ntsekouras , my bad.